### PR TITLE
Add support for object literal syntax in environment editor

### LINKED ISSUE #2762: Support Object Literals for environments instead of JSON
Since the application already converts the provided JSON object to a Object Literal. Not only would it be backwards-compat this change would reduce the time it takes to write the environment and also reduce the number of issues we see due to missing quotation marks as quotation marks inside of Object Literals is not required, but supported - as well as being able to support trailing commas.

Object Literal

```js
{
  firstName: "Xie",
  lastName: "Bo",
}
```

JSON

```json
{
  "firstName": "Xie",
  "lastName": "Bo"
}
```

### PR DESCRIPTION
**Pull request by diyorbek**
Closes #2762

I couldn't add tests because I couldn't figure out where I can write them. 
It would be great if I get some instructions about this.

### CODE REVIEW

File: packages/insomnia-app/app/ui/components/codemirror/lint/javascript-object-literal-lint.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// @flow
```
Comment (diyorbek) [inline packages/insomnia-app/app/ui/components/codemirror/lint/javascript-object-literal-lint.js:1]: Most part of this code is copied from `../javascript-async-lint.js`.
I was not sure if I could import `parseErrors` function from there.

File: packages/insomnia-app/app/ui/components/editors/environment-editor.js [hunk 2]
```diff
@@ -95,15 +97,19 @@ class EnvironmentEditor extends React.PureComponent<Props, State> {
 
   getValue(): EnvironmentInfo | null {
     if (this._editor) {
+      const textValue = this._editor.getValue();
+      const object = objectLiteralParser.parse(textValue);
+
       const data = orderedJSON.parse(
```
Comment (diyorbek) [inline packages/insomnia-app/app/ui/components/editors/environment-editor.js:103]: I didn't want to risk removing this part. Not sure if the `propertyOrder` still relevant.

Comment (develohpanda) [reply packages/insomnia-app/app/ui/components/editors/environment-editor.js:103]: It should be possible to remove the property order given the data is just stored as text now.

File: packages/insomnia-app/app/ui/components/modals/environment-edit-modal.js [hunk 3]
```diff
@@ -83,6 +87,15 @@ class EnvironmentEditModal extends PureComponent {
...
       propertyOrder: requestGroup && requestGroup.environmentPropertyOrder,
+      textValue: requestGroup
+        ? requestGroup.environmentTextValue === JSON.stringify({})
+          ? orderedJSON.stringify(
+              requestGroup.environment,
+              requestGroup.environmentPropertyOrder || null,
+              JSON_ORDER_SEPARATOR,
+            )
+          : requestGroup.environmentTextValue
+        : JSON.stringify({}),
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/components/modals/environment-edit-modal.js:98]: We can avoid this by writing a data migration in the `environment` model to ensure `environmentTextValue` is set correctly 🙂

**Review by develohpanda — COMMENTED** [review]
This is a really awesome PR! Generally it looks good to me, although I'll have to review it more thoroughly in the new year. Sorry about the delays! 🤗

File: packages/insomnia-app/app/ui/components/modals/environment-edit-modal.js (hunk 3 above)
Comment (diyorbek) [reply packages/insomnia-app/app/ui/components/modals/environment-edit-modal.js:98]: That would be great. Could you give me some directions on how to write a data migration? I couldn't quite get how it is done. Should it be done inside the `init` function of the model?

File: packages/insomnia-app/app/ui/components/editors/environment-editor.js (hunk 2 above)
Comment (diyorbek) [reply packages/insomnia-app/app/ui/components/editors/environment-editor.js:103]: Agree. I will try removing this part and test.

File: packages/insomnia-app/app/ui/components/modals/environment-edit-modal.js (hunk 3 above)
Comment (develohpanda) [reply packages/insomnia-app/app/ui/components/modals/environment-edit-modal.js:98]: Nope, it would be part of the `migrate` function.

Here's an example from the `workspace` model: https://github.com/Kong/insomnia/blob/6d5b8dd1545819b546073703545f4df2137daf95/packages/insomnia-app/app/models/workspace.js#L30-L35

File: packages/insomnia-app/app/ui/components/editors/environment-editor.js (hunk 2 above)
Comment (diyorbek) [reply packages/insomnia-app/app/ui/components/editors/environment-editor.js:103]: @develohpanda I think it's not 100% safe to remove `propertyOrder` because of [this PR](https://github.com/Kong/insomnia/pull/1863). 

I wrote the migration and removed `propertyOrder` from everywhere. Everything worked correctly.
Then I ran the demo provided in the PR above. It somehow worked fine. But the tests are failing as there is no `propertyOrder`.

Comment (diyorbek) [reply packages/insomnia-app/app/ui/components/editors/environment-editor.js:103]: Maybe we should rewrite those tests.
Or maybe there is something I missed while testing the demo manually.
I can't think of anything that can be an edge case in run time.

**Comment by jackkav** [comment]
Closed as stale.
